File: plugin/src/main/java/moe/hikari/canvas/variable/Variable.java
package moe.hikari.canvas.variable;

import org.jetbrains.annotations.Nullable;

import java.util.Set;

/**
 * 内存态变量记录。详见 {@code docs/dynamic-data.md §2.1}。
 *
 * <p>所有四类数据源（玩家变量 / 插件 push / 系统内置 / PAPI 桥接）统一用此 record 表达；
 * 持久化的只有 {@code namespace="user"} 的子集（落 {@code user_variables} 表）。</p>
 *
 * <p><b>fullName 编码</b>：
 * <ul>
 *   <li>普通 namespace：{@code <namespace>/<key>}（如 {@code bedwars/red_score}）</li>
 *   <li>user 变量：{@code user:<wallId>/<key>}（如 {@code user:w-3a17/红队比分}）——
 *       wallId 用冒号分隔避免 {@code user/<key>} 与 {@code user:<wall>/<key>} 冲突；
 *       对外 placeholder 文本里仍写 {@code ${var:user/红队比分}}，由 Compositor 注入 wallId。</li>
 * </ul></p>
 *
 * <p><b>线程安全</b>：record immutable；{@link #referencedByWalls} 由 {@link VariableStore}
 * 用 ConcurrentHashMap.Set view 维护，外部读取应视为不可变快照。</p>
 *
 * @param namespace          变量命名空间，{@code [a-zA-Z_][a-zA-Z0-9_:]*} ≤ 32（冒号给 user:<wallId> 用）
 * @param key                变量 key，{@code [a-zA-Z0-9_.-]+} ≤ 64
 * @param type               类型 hint
 * @param defaultValue       fallback 值；null = 无 fallback
 * @param currentValue       当前缓存值（push 写入 / 手动设 / Tier 3 算）
 * @param updatedAt          上次更新时间戳（unix ms）
 * @param ttl                TTL ms；0 = 永久；&gt;0 = 过期后渲染走 fallback 链
 * @param source             来源描述："manual" / "system" / 插件 plain name；null = 未声明
 * @param referencedByWalls  哪些 wall 引用该变量（倒排索引；Compositor 渲染前更新）
 */
public record Variable(
        String namespace,
        String key,
        VarType type,
        @Nullable String defaultValue,
        @Nullable String currentValue,
        long updatedAt,
        long ttl,
        @Nullable String source,
        Set<String> referencedByWalls
) {

    /**
     * 组合 fullName。普通 namespace = {@code <namespace>/<key>}；user 变量 namespace 已含
     * {@code user:<wallId>} 整段，本方法不知道也不需要拆。
     */
    public String fullName() {
        return namespace + "/" + key;
    }

    /** {@link #ttl} > 0 且 {@code now - updatedAt > ttl}。0 = 永久不过期。 */
    public boolean isStale(long now) {
        return ttl > 0 && (now - updatedAt) > ttl;
    }
}

File: web/src/types/variable.ts
/**
 * 变量系统前端类型（0.4.0-P1-D，对齐 {@code docs/dynamic-data.md §2 / §3} +
 * 后端 P1-A `moe.hikari.canvas.variable.Variable` record）。
 *
 * <p>核心 fullName 命名约定：</p>
 * <ul>
 *   <li>用户变量内部 namespace = {@code user:<wallId>}（冒号分隔 wallId），fullName 形如
 *     {@code user:w-3a17b2c1/红队比分}。这是后端 P1-A 固化的隔离方案，避免不同 wall 的
 *     user 变量名相撞。前端常用 {@link makeUserFullName} 拼。</li>
 *   <li>插件 / 系统 / PAPI 变量 namespace = {@code <pluginName>} / {@code system} /
 *     {@code papi}，fullName 用 {@link makeFullName} 直拼。</li>
 *   <li>{@code variable.create} op 的 {@code name} 字段**不含** {@code user:<wallId>/} 前缀，
 *     由后端注入。前端列表 / Picker 展示 fullName 时把 {@code user:<wallId>/} 段单独识别为
 *     "我的变量" 分组。</li>
 * </ul>
 *
 * <p>state.patch path 编码：variables 走 {@code /variables/<encoded>/...} 路径，
 * {@code encoded} 是 fullName 用 RFC 6901 {@code /} → {@code ~1} 转义后的字符串。前端
 * wsClient 解析时反向 unescape。</p>
 */

/** 变量值类型 hint；与后端 `VarType` enum 一一对应。 */
export type VarType = 'STRING' | 'NUMBER' | 'BOOLEAN' | 'COLOR';

/**
 * Variable 完整记录（state.patch add value / list 端点返回的形态）。
 *
 * - {@link namespace}：{@code user:<wallId>} / {@code userglobal}（0.4.3） / {@code system} /
 *   插件名 / {@code papi}
 * - {@link ttl}：毫秒；0 = 永久；>0 = TTL（最小 100ms 由后端校验）
 * - {@link source}：human-readable 来源；用户手动改 = {@code "manual"}；插件 push = 插件名；
 *   系统 provider = {@code "system"}；PAPI 桥接 = {@code "papi"}
 * - {@link updatedAt}：epoch ms
 * - {@link ownerUuid} / {@link ownerName}（0.4.3）：仅 {@code userglobal} namespace 有；其他 null
 */
export interface Variable {
    namespace: string;
    key: string;
    type: VarType;
    defaultValue: string | null;
    currentValue: string | null;
    updatedAt: number;
    ttl: number;
    source: string | null;
    ownerUuid?: string | null;
    ownerName?: string | null;
}

/** `variable.update` op 的 patch 字段；type / defaultValue 任一可省。 */
export interface VariablePatch {
    type?: VarType;
    defaultValue?: string | null;
}

/** Plugin Push API 批量 update 形态（P4 用，仅 ts 镜像供后续 typing 一致）。 */
export interface VariableUpdate {
    value: string;
    ttl?: number;
}

// ---------- fullName helpers ----------

/**
 * 给当前 wall 的 user 变量拼 fullName。例如
 * {@code makeUserFullName('w-3a17b2c1', '红队比分') → 'user:w-3a17b2c1/红队比分'}。
 */
export function makeUserFullName(wallId: string, key: string): string {
    return `user:${wallId}/${key}`;
}

/** 一般 fullName 拼接：{@code <namespace>/<key>}（不做转义）。 */
export function makeFullName(namespace: string, key: string): string {
    return `${namespace}/${key}`;
}

/**
 * 反向拆 fullName → namespace + key（按**首个** {@code /} 分割）。
 * 例如 {@code 'user:w-3a17b2c1/红队比分'} → {@code namespace: 'user:w-3a17b2c1'}, {@code key: '红队比分'}。
 * 无 {@code /} 时 namespace 为空、key 为原串。
 */
export function parseFullName(fullName: string): { namespace: string; key: string } {
    const idx = fullName.indexOf('/');
    if (idx < 0) return { namespace: '', key: fullName };
    return { namespace: fullName.substring(0, idx), key: fullName.substring(idx + 1) };
}

/** 判 namespace 是否为某 wall 的 user 变量（前缀 {@code user:}）。 */
export function isUserNamespace(namespace: string): boolean {
    return namespace.startsWith('user:');
}

/** 0.4.3：判 namespace 是否为全局用户变量（{@code userglobal}）。 */
export const USERGLOBAL_NAMESPACE = 'userglobal';
export function isUserGlobalNamespace(namespace: string): boolean {
    return namespace === USERGLOBAL_NAMESPACE;
}

/** 0.4.3：拼全局用户变量 fullName，例如 {@code 'userglobal/red_score'}。 */
export function makeUserGlobalFullName(key: string): string {
    return `${USERGLOBAL_NAMESPACE}/${key}`;
}
